server/src/main/java/org/elasticsearch/action/bulk/BulkRequestBuilder.java

```diff
@@ -126,32 +159,107 @@ public BulkRequestBuilder setWaitForActiveShards(final int waitForActiveShards)
-        request.routing(globalRouting);
+        this.globalRouting = globalRouting;
         return this;
     }
+
+    @Override
+    public BulkRequestBuilder setRefreshPolicy(WriteRequest.RefreshPolicy refreshPolicy) {
+        this.refreshPolicy = refreshPolicy;
+        return this;
+    }
+
+    @Override
+    public BulkRequestBuilder setRefreshPolicy(String refreshPolicy) {
+        this.refreshPolicyString = refreshPolicy;
+        return this;
+    }
+
+    @Override
+    public BulkRequest request() {
+        validate();
+        BulkRequest request = new BulkRequest(globalIndex);
+        if (requests.isEmpty() == false && requestBuilders.isEmpty() == false) {
+            throw new IllegalStateException("Must use only requests or request builders within a single bulk request");
+        }
+        for (RequestBuilder<? extends ActionRequest, ? extends ActionResponse> requestBuilder : requestBuilders) {
+            ActionRequest childRequest = requestBuilder.request();
+            request.add((DocWriteRequest<?>) childRequest);
+        }
+        for (DocWriteRequest<?> childRequest : requests) {
+            request.add(childRequest);
+        }
+        for (FramedData framedData : framedDataList) {
+            try {
+                request.add(framedData.data, framedData.from, framedData.length, framedData.defaultIndex, framedData.xContentType);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        }
+        if (waitForActiveShards != null) {
+            request.waitForActiveShards(waitForActiveShards);
+        }
+        if (timeout != null) {
+            request.timeout(timeout);
+        }
+        if (timeoutString != null) {
+            request.timeout(timeoutString);
+        }
+        if (globalPipeline != null) {
+            request.pipeline(globalPipeline);
+        }
+        if (globalRouting != null) {
+            request.routing(globalRouting);
+        }
+        if (refreshPolicy != null) {
+            request.setRefreshPolicy(refreshPolicy);
+        }
+        if (refreshPolicyString != null) {
+            request.setRefreshPolicy(refreshPolicyString);
+        }
+        return request;
+    }
+
+    private void validate() {
+        if (requests.isEmpty() == false && requestBuilders.isEmpty() == false
+            || requests.isEmpty() == false && framedDataList.isEmpty() == false
+            || requestBuilders.isEmpty() == false && framedDataList.isEmpty() == false) {
+            throw new IllegalStateException(
+                "Must use only request builders, requests, or byte arrays within a single bulk request. Cannot mix and match"
+            );
+        }
+        if (timeoutString != null && timeout != null) {
+            throw new IllegalStateException("Must use only one setTimeout method");
+        }
+        if (refreshPolicy != null && refreshPolicyString != null) {
```

Comment:
Low value contribution, but let the record show I was here: maybe these two `!= null` checks could be switch so they're symmetrical with the previous ones for the timeout handling.